### PR TITLE
Nmo13/features/multi selector

### PR DESCRIPTION
**Pull request by NMO13**
New timeline layout + support for multiple (up to 4) timelines

image

### CODE REVIEW

**Comment by NMO13** [comment]
- images are not visible any more
- multiselection
- crop timeline
- change cursor when over label

**Comment by thinkh** [comment]
- Images are working for me
- I fixed the cursor and style over timeline label

File: src/timeline/Timeline.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import {extractEpochId} from '../utils';
+import {DataStoreEpochSelection} from '../DataStore';
+import * as events from 'phovea_core/src/event';
+
+class SingleEpochSelector {
+  public $node: d3.Selection<any>;
+  hidden = true;
+  public curPos = -1;
+  constructor($parent: d3.Selection<any>, offsetH: number) {
+    this.$node = $parent.append('rect').classed('single-epoch-selector', true).attr('width', 2).attr('height', 30).attr('y', 0)
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  node(): d3.Selection<any> {
+    return this.$node;
+  }
+
+  render($parent, offsetH: number, offsetV: number, otl: OverallTimeline) {
+    this.build($parent);
+    this.$node.attr('transform', 'translate(0,' + offsetV + ')');
+
+    offsetH += 10;
+    const width = otl.dataPoints.length * 5;
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    const x = d3.scale.ordinal()
+      .rangePoints([0, width])
+      .domain(otl.dataPoints.map(function (d) {
+        return String(d.name);
+      }));
+
+    const xAxis = d3.svg.axis()
+      .scale(x)
+      .tickValues(x.domain().filter((d, i) => {
+        const cond = i < this.data.datapoints.length && this.data.datapoints[i].exists && i % 5 === 0;
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        return cond;
+      }))
+      .orient('bottom')
+      .tickSize(-7);
+
+    this.$node.append('g')
+      .attr('class', 'axis')
+      .attr('transform', `translate(${offsetH}, 15)`)
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    const $brushg = this.$node.append('g')
+      .attr('transform', `translate(${offsetH}, 0)`)
+      .attr('class', 'brush')
+      .call(brush);
+
+    brush.on('brush', () => this.brushmove(x, brush))
+         .on('brushend', () => that.brushend(x, brush));
+
+    $brushg.selectAll('rect')
+      .attr('height', 15);
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    };
+
+    const $singleSelectionArea = this.$node.append('rect').style('fill', 'rgb(0,0.255').attr('width', 2).attr('height', 15).attr('y', 0)
+      .attr('transform', `translate(${offsetH}, 0)`);
+
+    const tml = this;
+    this.singleEpochSelector = new SingleEpochSelector(this.$node, offsetH);
+    this.$node.append('rect').attr('transform', `translate(${offsetH}, ${16})`)
+      .attr('width', width)
+      .attr('height', 15)
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+      let pos = invert(coordinates[0]);
+      pos = Math.round(pos);
+      return pos;
+    };
+
+    const isValidPos = (pos: number) => {
+      return pos < this.data.datapoints.length && this.data.datapoints[pos].exists;
+    };
+
+    const $singleSelectionArea = this.$node.append('rect').style('fill', 'rgb(0,0.255').attr('width', 2).attr('height', 15).attr('y', 0)
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Extract constants

File: src/timeline/Timeline.ts [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      .attr('transform', `translate(${offsetH}, 0)`);
+
+    const tml = this;
+    this.singleEpochSelector = new SingleEpochSelector(this.$node, offsetH);
+    this.$node.append('rect').attr('transform', `translate(${offsetH}, ${16})`)
```
Comment (thinkh) [inline src/timeline/Timeline.ts]: Remove `${16}` or extract to constants

File: src/timeline/TimelineView.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  private width: number;
+
+  constructor(parent: Element) {
+    this.width = parent.clientWidth;
+    this.$node = d3.select(parent)
+      .append('svg')
+      .classed('timeline-view', true)
+      .attr('width', '100%')
+      .attr('height', '0px')
+      .attr('viewBox', `0 0 ${this.width} ${0}`);
```
Comment (thinkh) [inline src/timeline/TimelineView.ts]: Remove `${0}`

**Comment by thinkh** [comment]
@NMO13 Select the first dataset by default.

**Pull request merged**